packages/http-server/src/http-server.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/http-server
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {createServer, Server, ServerRequest, ServerResponse} from 'http';
+import {AddressInfo} from 'net';
+import * as pEvent from 'p-event';
+
+export type HttpRequestListener = (
+  req: ServerRequest,
+  res: ServerResponse,
+) => void;
+
+/**
+ * Object for specifyig the HTTP / HTTPS server options
+ */
+export type HttpServerOptions = {
+  port?: number;
+  host?: string;
+  protocol?: HttpProtocol;
+};
+
+export type HttpProtocol = 'http' | 'https'; // Will be extended to `http2` in the future
+
+/**
+ * HTTP / HTTPS server used by LoopBack's RestServer
+ *
+ * @export
+ * @class HttpServer
+ */
+export class HttpServer {
+  private _port: number;
+  private _host?: string;
+  /**
+   * Protocol, default to `http`
+   */
+  private _protocol: HttpProtocol;
+  private _address: AddressInfo;
+  private httpRequestListener: HttpRequestListener;
+  private httpServer: Server;
+
+  /**
+   * @param httpServerOptions
+   * @param httpRequestListener
+   */
+  constructor(
+    httpRequestListener: HttpRequestListener,
+    httpServerOptions?: HttpServerOptions,
+  ) {
+    this.httpRequestListener = httpRequestListener;
+    if (!httpServerOptions) httpServerOptions = {};
+    this._port = httpServerOptions.port || 0;
+    this._host = httpServerOptions.host || undefined;
+    this._protocol = httpServerOptions.protocol || 'http';
+  }
+
+  /**
+   * Starts the HTTP / HTTPS server
+   */
+  public async start() {
+    this.httpServer = createServer(this.httpRequestListener);
+    this.httpServer.listen(this._port, this._host);
+    await pEvent(this.httpServer, 'listening');
+    this._address = this.httpServer.address() as AddressInfo;
+    this._host = this._host || this._address.address;
+    this._port = this._address.port;
+  }
+
+  /**
+   * Stops the HTTP / HTTPS server
+   */
+  public async stop() {
+    if (this.httpServer) {
+      this.httpServer.close();
+      await pEvent(this.httpServer, 'close');
+    }
+  }
+
+  /**
+   * Protocol of the HTTP / HTTPS server
+   */
+  public get protocol(): HttpProtocol {
+    return this._protocol;
+  }
+
+  /**
+   * Port number of the HTTP / HTTPS server
+   */
+  public get port(): number {
+    return this._port;
+  }
+
+  /**
+   * Host of the HTTP / HTTPS server
+   */
+  public get host(): string | undefined {
+    return this._host;
+  }
+
+  /**
+   * URL of the HTTP / HTTPS server
+   */
+  public get url(): string {
+    return `${this._protocol}://${this.host}:${this.port}`;
```

Comment:
Please make it more robust per https://github.com/strongloop/loopback-next/blob/2bddc47ccda56a50c52a8a92facd87e3a3f4944d/packages/http-server/src/common.ts#L176-L201

Reply:
+1 to handle IPv6 localhost URLs correctly, although I am fine with leaving that out of scope of this pull request. Isn't there any Node.js core API or perhaps an npm module that can solve this problem for us?

The built-in `url` API seems to work well

```
$ node
> require('url').format({protocol: 'http', hostname: '::1', port: 80, pathname: '/foo'})
'http://[::1]:80/foo'
```

The docs says this code is considered as legacy and should not be used in new code :(

Unfortunately the newer WhatWG API does not encode IPv6 local addressed as the old one, so I am not sure how much benefit we can get from using it.

See https://github.com/nodejs/help/issues/1176#issuecomment-375131204

```ts
const u = new (require('url').URL)('http://./');
u.protocol = 'https';
u.host = '[::1]';
u.port = 3000;
u.pathname = '/foo';
u.href 
// returns 
// 'https://[::1]:3000/foo'
```

---

-1 for Travis-specific hacks, see https://github.com/strongloop/loopback-next/pull/1369#discussion_r192014280

> We cannot ship workarounds for Travis CI in code that will be used by LB apps in production!
>
> If IPv6 localhost addresses are causing problems in Travis, then let's fix our tests to explicitly listen on 127.0.0.1 instead.

Reply:
@raymondfeng can we do the "robustification" after landing the basic functionality?

Reply:
If we force tests to only listen `127.0.0.1`, we then lose the coverage for `host === undefined`.

Reply:
Makes sense. I am proposing to write two tests for `host === undefined` case:

 - One test to verify what is returned by `host` and `url` properties. I think we don't need to start the server for this and thus these tests can be at unit-test level.
 - One test to verify that server listening on all hosts accepts connections from `127.0.1`. The trick is to build the request URL manually (`http://127.0.0.1:${server.port}/`) in the test, don't use `server.url` property.